ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -373,6 +383,54 @@ void fillChunks() {
   }
 
   /**
+   * Submits encrypt job for the given {@link PutChunk} and processes the callback for the same
+   * @param putChunk the {@link PutChunk} that needs to be encrypted
+   * @param encryptContent {@code true} if content needs to be encrypted. {@code false} otherwise
+   */
+  private void encryptChunk(PutChunk putChunk, boolean encryptContent) {
+    try {
+      logger.trace("Chunk at index {} moves to {} state", putChunk.chunkIndex, ChunkState.Encrypting);
+      putChunk.state = ChunkState.Encrypting;
+      putChunk.chunkEncryptReadyAtMs = time.milliseconds();
+      logger.trace("Submitting encrypt job for chunk at index {}", putChunk.chunkIndex);
+      cryptoJobHandler.submitJob(
+          new EncryptJob(passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
+              encryptContent ? putChunk.buf : null, ByteBuffer.wrap(putChunk.chunkUserMetadata), kms.getRandomKey(),
+              cryptoService, kms, (EncryptJob.EncryptJobResult result, Exception exception) -> {
+            logger.trace("Processing encrypt job callback for chunk at index " + putChunk.chunkIndex);
+            if (exception == null && !isOperationComplete()) {
+              if (encryptContent) {
+                putChunk.buf = result.getEncryptedBlobContent();
+              }
+              putChunk.encryptedPerBlobKey = result.getEncryptedKey();
+              putChunk.chunkUserMetadata = result.getEncryptedUserMetadata().array();
```

Comment:
please write a method to get this safely

Reply:
sorry, may I know which one are you talking about. Do you mean to say, the entire callback processing ?

Reply:
No the `.array()`

Reply:
IIRC, the `array()` method does not work on read-only byte buffers.

Reply:
@vgkholla : userMetada is sent in as byte array to the router. It is the router that wraps it in a ByteBuf and sends it to storage. Do you think we need to be cautious in doing this.